log/format.go

```diff
@@ -98,212 +73,102 @@ type TerminalStringer interface {
-			fieldPadding[k] = padding
+			fieldPadding[key] = padding
 			fieldPaddingLock.Unlock()
 		}
 		if color > 0 {
-			fmt.Fprintf(buf, "\x1b[%dm%s\x1b[0m=", color, k)
+			fmt.Fprintf(buf, "\x1b[%dm%s\x1b[0m=", color, key)
 		} else {
-			buf.WriteString(k)
+			buf.WriteString(key)
 			buf.WriteByte('=')
 		}
-		buf.WriteString(v)
-		if i < len(ctx)-2 && padding > length {
+		buf.WriteString(val)
+		if i < r.NumAttrs()-1 && padding > length {
 			buf.Write(bytes.Repeat([]byte{' '}, padding-length))
 		}
 	}
 	buf.WriteByte('\n')
 }
 
-// JSONFormat formats log records as JSON objects separated by newlines.
-// It is the equivalent of JSONFormatEx(false, true).
-func JSONFormat() Format {
-	return JSONFormatEx(false, true)
-}
-
-// JSONFormatOrderedEx formats log records as JSON arrays. If pretty is true,
-// records will be pretty-printed. If lineSeparated is true, records
-// will be logged with a new line between each record.
-func JSONFormatOrderedEx(pretty, lineSeparated bool) Format {
-	jsonMarshal := json.Marshal
-	if pretty {
-		jsonMarshal = func(v interface{}) ([]byte, error) {
-			return json.MarshalIndent(v, "", "    ")
-		}
-	}
-	return FormatFunc(func(r *Record) []byte {
-		props := map[string]interface{}{
-			r.KeyNames.Time: r.Time,
-			r.KeyNames.Lvl:  r.Lvl.String(),
-			r.KeyNames.Msg:  r.Msg,
-		}
-
-		ctx := make([]string, len(r.Ctx))
-		for i := 0; i < len(r.Ctx); i += 2 {
-			if k, ok := r.Ctx[i].(string); ok {
-				ctx[i] = k
-				ctx[i+1] = formatLogfmtValue(r.Ctx[i+1], true)
-			} else {
-				props[errorKey] = fmt.Sprintf("%+T is not a string key,", r.Ctx[i])
-			}
-		}
-		props[r.KeyNames.Ctx] = ctx
-
-		b, err := jsonMarshal(props)
-		if err != nil {
-			b, _ = jsonMarshal(map[string]string{
-				errorKey: err.Error(),
-			})
-			return b
-		}
-		if lineSeparated {
-			b = append(b, '\n')
-		}
-		return b
-	})
-}
-
-// JSONFormatEx formats log records as JSON objects. If pretty is true,
-// records will be pretty-printed. If lineSeparated is true, records
-// will be logged with a new line between each record.
-func JSONFormatEx(pretty, lineSeparated bool) Format {
-	jsonMarshal := json.Marshal
-	if pretty {
-		jsonMarshal = func(v interface{}) ([]byte, error) {
-			return json.MarshalIndent(v, "", "    ")
-		}
-	}
-
-	return FormatFunc(func(r *Record) []byte {
-		props := map[string]interface{}{
-			r.KeyNames.Time: r.Time,
-			r.KeyNames.Lvl:  r.Lvl.String(),
-			r.KeyNames.Msg:  r.Msg,
-		}
-
-		for i := 0; i < len(r.Ctx); i += 2 {
-			k, ok := r.Ctx[i].(string)
-			if !ok {
-				props[errorKey] = fmt.Sprintf("%+T is not a string key", r.Ctx[i])
-			} else {
-				props[k] = formatJSONValue(r.Ctx[i+1])
-			}
-		}
-
-		b, err := jsonMarshal(props)
-		if err != nil {
-			b, _ = jsonMarshal(map[string]string{
-				errorKey: err.Error(),
-			})
-			return b
-		}
-
-		if lineSeparated {
-			b = append(b, '\n')
-		}
-
-		return b
-	})
-}
-
 func formatShared(value interface{}) (result interface{}) {
```

Comment:
This `formatShared` method is now somewhat moot. I tried to hit it, with the following testcase

```golang
type customError struct{}

func (c *customError) Error() string { return "" }

type customStruct struct {
	A string
}

func TestLogging(t *testing.T) {
	b := new(bytes.Buffer)
	logger := NewLogger(NewGlogHandler(TerminalHandler(b, false)))

	logger.Info("test",
		"nil-stringer-interface", (fmt.Stringer)(nil),
		"nil-error-interface", error(nil),
		"nil-custom-err", (*customError)(nil),
		"nil-concrete-stringer", (*time.Time)(nil),
		"nil-custom-struct", (*customStruct)(nil),
	)
	t.Logf(b.String())
}

```

- The `case time.Time` was never hit (handled earlier in formatLogfmtValue)
- The `nil-concrete-stringer` is the only one that triggers a panic, and where we need the `recover`. 

Also there's this wonky case of `nil` vs `<nil>` (not  your fault, been like this since a long way back): 
```
    logger_test.go:50: INFO [11-09|08:14:53.459] test                                     nil-stringer-interface=nil nil-error-interface=nil nil-custom-err= nil-concrete-stringer=nil nil-custom-struct=<nil>
```

Reply:
It used to be that `formatShared` was called either from `formatJSONValue` or `formatLogfmtValue`. Now the former is removed, we can inline it into `formatLogfmtValue`. That makes a lot ot sense: both `formatLogfmtValue` and `formatShared` are just type-switches on `v`, so it's a lot better to merge them.

Reply:
As far as I can make out, the `formatShared` can essentially be reduced into 

```golang
func formatShared(value interface{}) (result interface{}) {
	switch v := value.(type) {
	case error:
		return v.Error()

	case fmt.Stringer:
		if v := reflect.ValueOf(v); v.Kind() == reflect.Ptr && v.IsNil() {
			return "caughtnil"
		}
		return v.String()
	default:
		return v
	}
}
```
(and then inlined above)